### ISSUE TITLE
Using cmake with an alternate toolchain

### ISSUE THREAD

**Issue by dbry** [dbry/WavPack#146]
Since there have been numerous changes to the cmake stuff here, including the ability to build the winamp plugin for Window using mingw, I thought I would experiment with this (if for no other reason to test the generated winamp plugin and see if it actually works). I have mingw installed and use it to generate Windows executables all the time (by itself, or with Autotools to build WavPack), so I thought this would be fairly straightforward.

Initially I had no idea how to start so I Googled and found [this page](https://gist.github.com/peterspackman/8cf73f7f12ba270aa8192d6911972fe8). I tried that and got pretty far, but it fails with both mingw versions here:
```
Performing C SOURCE FILE Test COMPILER_SUPPORTS_SYMBOL_MAPS failed with the following compile output:
Change Dir: /home/david/WavPack/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_be63c/fast && /usr/bin/make -f CMakeFiles/cmTC_be63c.dir/build.make CMakeFiles/cmTC_be63c.dir/build
make[1]: Entering directory '/home/david/WavPack/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_be63c.dir/src.c.obj
/usr/bin/x86_64-w64-mingw32-gcc   -DCOMPILER_SUPPORTS_SYMBOL_MAPS -Wl,--version-script=/home/david/WavPack/build//CMakeFiles/conftest.map   -o CMakeFiles/cmTC_be63c.dir/src.c.obj   -c /home/david/WavPack/build/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_be63c.exe
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_be63c.dir/link.txt --verbose=1
/usr/bin/cmake -E remove -f CMakeFiles/cmTC_be63c.dir/objects.a
/usr/bin/x86_64-w64-mingw32-ar cr CMakeFiles/cmTC_be63c.dir/objects.a @CMakeFiles/cmTC_be63c.dir/objects1.rsp
/usr/bin/x86_64-w64-mingw32-gcc  -DCOMPILER_SUPPORTS_SYMBOL_MAPS -Wl,--version-script=/home/david/WavPack/build//CMakeFiles/conftest.map    -Wl,--whole-archive CMakeFiles/cmTC_be63c.dir/objects.a -Wl,--no-whole-archive  -o cmTC_be63c.exe -Wl,--out-implib,libcmTC_be63c.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles/cmTC_be63c.dir/linklibs.rsp
/usr/bin/x86_64-w64-mingw32-ld:/home/david/WavPack/build//CMakeFiles/conftest.map:8: ignoring invalid character `\' in script
make[1]: Leaving directory '/home/david/WavPack/build/CMakeFiles/CMakeTmp'
```
Figuring Google would rescue me again I Googled `COMPILER_SUPPORTS_SYMBOL_MAPS` but found to my surprise that the only place on the whole Internet where that string exists is **my** `CMakeLists.txt`, which of course means I'm stuck.

Could anyone kindly put me on the right track?  @evpobr @sezero 

Also, is there an easy way to switch to clang?

Thanks!

**Comment by evpobr** [comment]
https://github.com/dbry/WavPack/blob/ce013734a99041b9cd9755428612ff9b7f6bb24d/CMakeLists.txt#L353

It checks `-Wl,--version-script` linker flag. I suspect that MinGW GCC does not support this.

Actually correct path is:

https://github.com/dbry/WavPack/blob/ce013734a99041b9cd9755428612ff9b7f6bb24d/CMakeLists.txt#L330

I am almost one hundred percent sure that this should work with MinGW on Windows.

**Comment by evpobr** [comment]
> /usr/bin/x86_64-w64-mingw32-ld:/home/david/WavPack/build//CMakeFiles/conftest.map:8: ignoring invalid character `\' in script

This is also very strange.

**Comment by evpobr** [comment]
```
Run Build Command(s):/usr/bin/make
```

Also this is strange. Not sure it is correct version. Maybe something is not installed.

**Comment by evpobr** [comment]
Probably generator must be `MinGW Makefiles`.

**Comment by sezero** [comment]
>   found [this page](https://gist.github.com/peterspackman/8cf73f7f12ba270aa8192d6911972fe8).

For what it's worth, this is my toolchain file for cross-compilation purposes:
```
SET(CMAKE_SYSTEM_NAME Windows)

SET(CMAKE_C_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-gcc)
SET(CMAKE_CXX_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-g++)
SET(CMAKE_RC_COMPILER /opt/cross_win64/bin/x86_64-w64-mingw32-windres)

# where is the target environment
SET(CMAKE_FIND_ROOT_PATH /opt/cross_win64 /opt/cross_win64/x86_64-w64-mingw32)

# adjust the default behaviour of the FIND_XXX() commands:
# search headers and libraries in the target environment, search
# programs in the host environment
SET(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
SET(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
SET(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

> https://github.com/dbry/WavPack/blob/ce013734a99041b9cd9755428612ff9b7f6bb24d/CMakeLists.txt#L353
> 
> It checks `-Wl,--version-script` linker flag. I suspect that MinGW GCC does not support this.

It actually should, if I remember correctly

[snipping other stuff]

> Could anyone kindly put me on the right track?

I am mentioned, but I don't speak and not proficient in cmake, I don't
like cmake, I detest cmake, [append more here ...]

I don't know whether he'd be interested in, but @madebr did quite a lot
of cmake work in SDL projects, and he _is_ proficient.

**Comment by madebr** [comment]
The error is about an invalid stray `\` in `conftest.map`
```
/usr/bin/x86_64-w64-mingw32-ld:/home/david/WavPack/build//CMakeFiles/conftest.map:8: ignoring invalid character `\' in script
```

This file is generated here:
https://github.com/dbry/WavPack/blob/ce013734a99041b9cd9755428612ff9b7f6bb24d/CMakeLists.txt#L351-L352

I ran these 2 lines locally, and it generated
```
VERS_1 {
    global: sym\;

};

VERS_2 {
    global: sym;
} VERS_1\;
```
which looks incorrect.

The string not escape the semicolons.
When fixing this, perhaps also add a final newline.

**Comment by madebr** [comment]
No idea why that's happening.
I just downloaded [a CMake 3.9.6 binary](https://cmake.org/files/v3.9/cmake-3.9.6-Darwin-x86_64.tar.gz),
created `/tmp/a.cmake` with:
```cmake
set(CONFTTEST_CONTENTS "VERS_1 {\n    global: sym\;\n\n};\n\nVERS_2 {\n    global: sym;\n} VERS_1\;")
file(WRITE conftest.map "${CONFTTEST_CONTENTS}")
```
And ran:
```
cmake -P /tmp/a.cmake
```
Both 3.9.6 and my system CMake 3.24.2 generated the same `conftest.map`.

**Comment by sezero** [comment]
In any case, removal of escaping semicolons _should_ not hurt.

**Comment by madebr** [comment]
No, the escape is only needed when you need to embed a semicolon as member of a list.

**Comment by sezero** [comment]
So, the following is a go? (I couldn't make it to emit the missing newline to end, though.)
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 597efac..5bd290f 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -350,3 +350,3 @@ elseif(CMAKE_SYSTEM_NAME MATCHES "Darwin")
 else()
-    set(CONFTTEST_CONTENTS "VERS_1 {\n    global: sym\;\n\n};\n\nVERS_2 {\n    global: sym;\n} VERS_1\;")
+    set(CONFTTEST_CONTENTS "VERS_1 {\n    global: sym;\n\n};\n\nVERS_2 {\n    global: sym;\n} VERS_1;")
     file(WRITE ${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/conftest.map "${CONFTTEST_CONTENTS}")
```

**Comment by sezero** [comment]
Off-topic to this ticket: would the existing wavpack cmake'ry work when embedding in other projects? I.e.: I might cook wavpack support to our SDL_mixer (hopefully soon (TM)).

**Comment by dbry** [comment]
I believe that embedding in other projects was part of the original intent. WavPack was recently added to Audacity, and I believe that's how they did it (but really have only a passing knowledge).

On-topic, the fix above eliminates the warning message (which occurs with regular CC also). However, apparently it wasn't the cause of the error return. It still fails there, this time with no syntax warning message:
```
Performing C SOURCE FILE Test COMPILER_SUPPORTS_SYMBOL_MAPS failed with the following compile output:
Change Dir: /home/david/WavPack/build/CMakeFiles/CMakeTmp

Run Build Command(s):/usr/bin/make cmTC_48e24/fast && /usr/bin/make -f CMakeFiles/cmTC_48e24.dir/build.make CMakeFiles/cmTC_48e24.dir/build
make[1]: Entering directory '/home/david/WavPack/build/CMakeFiles/CMakeTmp'
Building C object CMakeFiles/cmTC_48e24.dir/src.c.obj
/usr/bin/i686-w64-mingw32-gcc   -DCOMPILER_SUPPORTS_SYMBOL_MAPS -Wl,--version-script=/home/david/WavPack/build//CMakeFiles/conftest.map   -o CMakeFiles/cmTC_48e24.dir/src.c.obj   -c /home/david/WavPack/build/CMakeFiles/CMakeTmp/src.c
Linking C executable cmTC_48e24.exe
/usr/bin/cmake -E cmake_link_script CMakeFiles/cmTC_48e24.dir/link.txt --verbose=1
/usr/bin/cmake -E remove -f CMakeFiles/cmTC_48e24.dir/objects.a
/usr/bin/i686-w64-mingw32-ar cr CMakeFiles/cmTC_48e24.dir/objects.a @CMakeFiles/cmTC_48e24.dir/objects1.rsp
/usr/bin/i686-w64-mingw32-gcc  -DCOMPILER_SUPPORTS_SYMBOL_MAPS -Wl,--version-script=/home/david/WavPack/build//CMakeFiles/conftest.map    -Wl,--whole-archive CMakeFiles/cmTC_48e24.dir/objects.a -Wl,--no-whole-archive  -o cmTC_48e24.exe -Wl,--out-implib,libcmTC_48e24.dll.a -Wl,--major-image-version,0,--minor-image-version,0 @CMakeFiles/cmTC_48e24.dir/linklibs.rsp
make[1]: Leaving directory '/home/david/WavPack/build/CMakeFiles/CMakeTmp'
```
So @sezero , I assume this step passes in your setup? I thought this would be easy considering that you were doing basically the same thing, but it's certainly not critical to what I'm doing (I still use Autotools for day-to-day work).

Thanks @madebr and @evpobr !

**Comment by madebr** [comment]
`FILE_CONTENTS` is also handled weird. It's a string, which is appended as a string.
CMake has `string(APPEND)`, but not in cmake 3.2.

I'll fork and take a looksy.

**Comment by madebr** [comment]
With https://github.com/dbry/WavPack/pull/147, I can build wavpack targeting Linux and MinGW, on my Linux machine.

**Comment by dbry** [comment]
Me too, including the winamp and cooledit plugins. Thanks so much!

**Comment by sezero** [comment]
> One thing to note here: COMPILER_SUPPORTS_SYMBOL_MAPS needn't (shouldn't) be checked at all if we are not building a shared lib. I believe that should be rectified.

Unless the existing behavior is intentional (I can't imagine why that would be),
is the following patch draft good?
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index b5f5cee..9d1d5f5 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -328,7 +328,8 @@ set(WAVPACK_EXPORT_SYMBOLS
     WavpackWriteTag
 )
 
-if(BUILD_SHARED_LIBS AND (WIN32 OR CYGWIN))
+if(BUILD_SHARED_LIBS)
+  if(WIN32 OR CYGWIN)
     set(FILE_CONTENTS "EXPORTS\n")
     foreach(EXPORT_SYMBOL ${WAVPACK_EXPORT_SYMBOLS})
         list(APPEND FILE_CONTENTS "    ${EXPORT_SYMBOL}\n")
@@ -336,8 +337,7 @@ if(BUILD_SHARED_LIBS AND (WIN32 OR CYGWIN))
 
     file(WRITE ${CMAKE_CURRENT_BINARY_DIR}/libwavpack.def ${FILE_CONTENTS})
     target_sources(wavpack PRIVATE ${CMAKE_CURRENT_BINARY_DIR}/libwavpack.def)
-
-elseif(CMAKE_SYSTEM_NAME MATCHES "Darwin")
+  elseif(CMAKE_SYSTEM_NAME MATCHES "Darwin")
     set(FILE_CONTENTS "")
     foreach(EXPORT_SYMBOL ${WAVPACK_EXPORT_SYMBOLS})
         list(APPEND FILE_CONTENTS "_${EXPORT_SYMBOL}\n")
@@ -348,7 +348,7 @@ elseif(CMAKE_SYSTEM_NAME MATCHES "Darwin")
     else()
         target_link_directories(wavpack PRIVATE "-Wl,-exported_symbols_list,${CMAKE_CURRENT_BINARY_DIR}/libwavpack.sym")
     endif()
-else()
+  else()
     set(CONFTTEST_CONTENTS "VERS_1 {\n    global: sym;\n};\n\nVERS_2 {\n    global: sym;\n} VERS_1;\n")
     file(WRITE ${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/conftest.map "${CONFTTEST_CONTENTS}")
     check_c_linker_flag("-Wl,--version-script=${PROJECT_BINARY_DIR}/${CMAKE_FILES_DIRECTORY}/conftest.map" COMPILER_SUPPORTS_SYMBOL_MAPS)
@@ -366,6 +366,7 @@ else()
             target_link_options(wavpack PRIVATE "-Wl,--version-script=${CMAKE_CURRENT_BINARY_DIR}/libwavpack.map;-Wl,-no-undefined")
         endif()
     endif()
+  endif()
 endif()
 
 set_target_properties(wavpack PROPERTIES EXPORT_NAME WavPack)
```

**Comment by madebr** [comment]
That sounds correct to me.

**Comment by sezero** [comment]
> That sounds correct to me.

@dbry: The patch is above, but if you want a PR I can create one.

**Comment by dbry** [comment]
@sezero That's all right...I can push the change, unless you want the credit...  :smiley_cat: 

The purpose of this exercise for me was to verify the operation of the plugins, and unfortunately I have found that neither of them work. I noticed when the cooledit filter is building I got a whole batch of these warnings which look to me like somehow the Pascal calling convention is not happening correctly:
```
/usr/bin/i686-w64-mingw32-ld: warning: resolving _CloseFilterInput by linking to _CloseFilterInput@4
Use --enable-stdcall-fixup to disable these warnings
Use --disable-stdcall-fixup to disable these fixups
/usr/bin/i686-w64-mingw32-ld: warning: resolving _CloseFilterOutput by linking to _CloseFilterOutput@4
/usr/bin/i686-w64-mingw32-ld: warning: resolving _FilterGetFileSize by linking to _FilterGetFileSize@4
...
```
When cooledit starts I get this message: 
![CooleditError-126](https://user-images.githubusercontent.com/5532227/202568815-7b030fc7-43bc-4add-ae6c-4f3fe7229bfc.png)



I don't get any warnings during the build step, but the winamp plugin fails to load also:
![WinampError](https://user-images.githubusercontent.com/5532227/202568819-556c83aa-fdc0-44be-bdb9-37629227c8df.png)

One difference on the winamp plugin is that the `file` report is slightly different. The working plugin displays
```
in_wv.dll: PE32 executable (DLL) (GUI) Intel 80386, for MS Windows
```
whereas the one built now shows:
```
in_wv.dll: PE32 executable (DLL) (console) Intel 80386, for MS Windows
```

Also, I believe that the `in_wv.lng` gets included in the dll because the .lng file is never referenced in the NSI installation script.

However, I mention this only for completeness, This is not at all important to me because both of these plugins are rather obsolete and I can still build them on MSVC if there's ever a need to do so again. I also don't have enough time or interest (or knowledge) to figure out what might be happening. I am actually more concerned about building non-functioning plugins that someone might try to use (although that's also unlikely). At a minimum I think I'll put a note somewhere that plugins built with cmake do not work.

**Comment by madebr** [comment]
I am also seeing these warnings.
These are about the name of the symbol names created in the `.c` files not matching up with the exported names in the `.def` file.

MinGW does `#define PASCAL __stdcall`, so these symbols are actually built with stdcall convention.

What is the difference when running  `dumpbin /exports` on the 2 dll's?

Instead of using a `.def` file and/or adding all these `/export:symbol1 /export:symbol2` in the `.vcxproj`,
I think it is easier to add the following to e.g. `audition/filters.h`
```c
#if defined(AUDITON_EXPORTS)
#define AUDITION_EXTERN __declspec(dllexport)
#else
#define AUDITION_EXTERN
#endif
```

and use it as:

```c
AUDITION_EXTERN short PASCAL QueryCoolFilter (COOLQUERY *lpcq);
```

**Comment by sezero** [comment]
The attached patch removes exported symbol lists from cmake and msvc project file, and marks exported syms explicitly with `__declspec(dllexport)` (Hopefully I didn't screw up the msvc project file..)

[patch_exports.txt](https://github.com/dbry/WavPack/files/10036363/patch_exports.txt)

**Comment by sezero** [comment]
> > Also, I believe that the `in_wv.lng` gets included in the dll because the .lng file is never referenced in the NSI installation script.
> 
> That in_wv.lng thing only builds in the resource, so I really don't understand its purpose. There is an MSVC project for it, and the MSVC project for the dll itself doesn't include the resource, which is weird. I don't know anything about winamp or its plugins so I can't be of help with it...

The lng file thing seems to be used this way:  https://github.com/dbry/WavPack/blob/master/winamp/in_wv.c#L85-L104

**Comment by dbry** [comment]
First of all, many thanks to @madebr and @sezero for your help here!

I have pushed the first two patches above and they all seem fine. I can now build working executables in `mingw` using `cmake`.

Unfortunately neither of the patches to fix the exports generates workable plugin files. I checked the exports of the mingw-generated file with `dumpbin` and now the symbols have the numeric decoration at the end (which the originals do not) so I knew they wouldn't work (but did try).

I also read [here](https://www.willus.com/mingw/yongweiwu_stdcall.html) that MSVC can change the decoration depending on whether the symbols come from a .def file or the __declspec(dllexport) attribute (what?). So maybe that's why the MSVC project does it the way it does, and in fact when I applied the second patch and built in MSVC the symbol names got the trailing numeric _and_ a preceding underscore added, and so again, that didn't work.

At this point I'm not even curious as to why this is so messed up, and am just happy that I can still build working plugins in MSVC!  :smiley:

**Comment by dbry** [comment]
> > > Also, I believe that the `in_wv.lng` gets included in the dll because the .lng file is never referenced in the NSI installation script.
> > 
> > 
> > That in_wv.lng thing only builds in the resource, so I really don't understand its purpose. There is an MSVC project for it, and the MSVC project for the dll itself doesn't include the resource, which is weird. I don't know anything about winamp or its plugins so I can't be of help with it...
> 
> The lng file thing seems to be used this way: https://github.com/dbry/WavPack/blob/master/winamp/in_wv.c#L85-L104

Yes, it's starting to come back to be (that was a loooong time ago). I think the `in_wv.lng` info is in the resource file and so it's in the DLL and that gets used to provide the text for the dialogs. But if there's a `in_wv.lng` file then there's an attempt to load alternate language strings for the dialogs from there (which is the code you see) that overrides the strings based on the language setting. The file generated can be used to test that facility, but normally it's not there. I don't know if that dialog was ever Internationalized, but if it was then that file would be present. Maybe. But anyway, it _doesn't_ get included with the plugin distribution.

**Comment by evpobr** [comment]
> ```c
> #if defined(AUDITON_EXPORTS)
> #define AUDITION_EXTERN __declspec(dllexport)
> #else
> #define AUDITION_EXTERN
> #endif
> ```

Actually you need define here and no ifdefs. You never build Cool Edit plugin as static library and nobody includes its header somewhere and link it using import library.

**Comment by evpobr** [comment]
I guess Cool Edit plugin uses `PASCAL` (`stdcall`) calling convension, but symbols must be undecorated. MS linker resolves this without problems, but not `ld`.

Found this snippet:

https://raw.githubusercontent.com/Microsoft/vcpkg/master/ports/libssh/0002-mingw_for_Android.patch

```cmake
if (MINGW AND NOT ANDROID)
     target_link_libraries(ssh PRIVATE "-Wl,--enable-stdcall-fixup")
     target_compile_definitions(ssh PRIVATE "_POSIX_SOURCE")
 endif ()
```

@dbry , just use native platform tools, eg Visual Studio for Windows.

And here:

https://sourceware.org/binutils/docs-2.18/ld/Options.html

> --enable-stdcall-fixup
--disable-stdcall-fixup
If the link finds a symbol that it cannot resolve, it will attempt to do “fuzzy linking” by looking for another defined symbol that differs only in the format of the symbol name (cdecl vs stdcall) and will resolve that symbol by linking to the match. For example, the undefined symbol _foo might be linked to the function _foo@12, or the undefined symbol _bar@16 might be linked to the function _bar. When the linker does this, it prints a warning, since it normally should have failed to link, but sometimes import libraries generated from third-party dlls may need this feature to be usable. If you specify --enable-stdcall-fixup, this feature is fully enabled and warnings are not printed. If you specify --disable-stdcall-fixup, this feature is disabled and such mismatches are considered to be errors. [This option is specific to the i386 PE targeted port of the linker]

**Comment by sezero** [comment]
> I also read [here](https://www.willus.com/mingw/yongweiwu_stdcall.html) that MSVC can change the decoration depending on whether the symbols come from a .def file or the __declspec(dllexport) attribute

Aghh.. (That's a link I lost long ago, thanks for pointing it, now bookmarked it.)
So, different solutions needed for different toolchains, and since cmake supports
different toolchains one solution adversely affects the other: decision is WONTFIX
then, yes?

Still curious though, as to why the winamp plugin doesn't work despite it being
not stdcall...

**Comment by evpobr** [comment]
And I also didn't quite understand - both WinAmp and Cool Edit plugins don't load either? Or did it work out with one of them?

**Comment by madebr** [comment]
I created #148 with little cmake fixes.

**Comment by dbry** [comment]
> And I also didn't quite understand - both WinAmp and Cool Edit plugins don't load either? Or did it work out with one of them?

Neither plugin worked. The `Cooledit` failure is probably related to the `PASCAL` calling sequence (which is basically `stdcall` with the arguments pushed in reverse order). The fact that mingw does `#define PASCAL __stdcall` is not encouraging.

In the `Winamp` case it could be related to all the C++ going on there. My understanding is that there's no well-defined ABI for C++ like there is for C and so interoperability is not a given. 

Another issue could be dependencies. Both MSVC-generated DLLs only depend on `USER32` and `KERNEL32`. Checking the mingw-generated versions I saw things like `libgcc_s_sj1j_1.dll` and `libstdc++-6.dll`. :astonished:

Definitely WONTFIX.

**Comment by sezero** [comment]
>   `Cooledit` failure is probably related to the `PASCAL` calling sequence (which is basically `stdcall` with the arguments pushed in reverse order). The fact that mingw does `#define PASCAL __stdcall` is not encouraging.

MSVC also defines `PASCAL` as `__stdcall`, there is no issue there.

**Comment by madebr** [comment]
I fixed the WinAMP plugin in https://github.com/dbry/WavPack/pull/149

**Comment by madebr** [comment]
About pascal vs stdcall, win16 used the pascal calling convention. win32 made the switch to stdcall.
See this [old new thing blog post talking about win16](https://learn.microsoft.com/en-us/cpp/cpp/obsolete-calling-conventions?view=msvc-170), and [this post talking about win32](https://devblogs.microsoft.com/oldnewthing/20040108-00/?p=41163)

**Comment by sezero** [comment]
#148 seems to have introduced a typo in the resource file. @dbry: you want the following:
```diff
diff --git a/wavpackdll/wavpackdll.rc b/wavpackdll/wavpackdll.rc
index 4ddaa3a..d75a807 100644
--- a/wavpackdll/wavpackdll.rc
+++ b/wavpackdll/wavpackdll.rc
@@ -35,5 +35,5 @@ END
 2 TEXTINCLUDE 
 BEGIN
-    "#include <winresrc.rc>\r\n"
+    "#include <winresrc.h>\r\n"
     "\0"
 END
```

**Comment by sezero** [comment]
I wonder though: After #149, the cmake `@` patch at https://github.com/dbry/WavPack/issues/146#issuecomment-1319309376 may now work? Both for cmake + mingw, and for cmake + msvc?

**Comment by evpobr** [comment]
I think WinaAmp plugin  is ready. Not sure about Cool Edit. I guess it requires something like #149. And probably `-enable-stdcall-fixup` fix or `.DEF` file.

I'm sure that everything will eventually work with MinGW and MSVC, but I wonder how exactly.

**Comment by sezero** [comment]
> I think WinaAmp plugin is ready. Not sure about Cool Edit. I guess it requires something like #149. And probably `-enable-stdcall-fixup` fix or `.DEF` file.
> 
> I'm sure that everything will eventually work with MinGW and MSVC, but I wonder how exactly.

Cmake already generates a def file for cooledit plugin. The patch at https://github.com/dbry/WavPack/issues/146#issuecomment-1319309376 just adds the missing `@nn` stdcall suffixes to it, and that's why I think that that would make it work with both with mingw and msvc.

**Comment by madebr** [comment]
@sezero The MSVC project does not set the `@XX` suffixes either.
I built all dll's with cmake and msvc and didn't see any difference when running `dumpbin /exports`.

@dbry 
Don't forget about https://github.com/dbry/WavPack/issues/146#issuecomment-1320852594.
With https://github.com/dbry/WavPack/commit/6b20158b392ea93dae03937311ef2fcc1d74e7a8, does the cooledit plugin work when built with CMake?

**Comment by sezero** [comment]
> @sezero The MSVC project does not set the `@XX` suffixes either.

Yes: I can only guess that `link.exe` does some magic behind doors to handle things.

> I built all dll's with cmake and msvc and didn't see any difference when running `dumpbin /exports`.

So, the existing cmake file really gives a proper dll when using msvc as the toolchain?
That's really unexpected.

**Comment by dbry** [comment]
> > @sezero The MSVC project does not set the `@XX` suffixes either.
> 
> > I built all dll's with cmake and msvc and didn't see any difference when running `dumpbin /exports`.
> 
> So, the existing cmake file really gives a proper dll when using msvc as the toolchain? That's really unexpected.

Here's the status. Using cmake and mingw I now get fully working Winamp and Cooledit plugins. The Cooledit symbols are completely undecorated (as required). I do get the warnings about the fixups we saw before, but those should be easy to silence.

Of course `MSVC` using project files still works. What I don't know how to do is try using `cmake` with `MSVC`. I assume that involves installing `cmake` for Windows and running it in a `MSVC` developer console? I could do that, but if @madebr has verified that the exported symbols are identical then it's probably good.

What I'm starting to understand is that even though the `PASCAL` keyword is used in the Cooledit plugin, it really is just `__stdcall` and that nobody has _really_ used `PASCAL` calling convention (where the arg push order is reversed) since the 16-bit days (kind of like `FAR`). It could just be replaced with `WINAPI` or `__stdcall`.

One thing I noticed when I tried the `Cooledit` filter was that it seemed slow, and so started looking into performance. What I see is that the `cmake` native builds are almost _twice as slow_ as `Autotools` builds and the `cmake` `mingw` builds are almost _three times slower_ (i686 or x86_64)! Under `Autotools` everything is about the same.

So, why? The first thing I noticed is that when using `mingw` with `cmake` the assembler optimizations are disabled whereas natively they're used. That could explain the extra slowdown when switching from native to `mingw` under `cmake`. But it doesn't explain the initial performance hit just switching to `cmake`. Is it possible that it's simply not using `-O2` or friend?

Thanks again guys...I'm forever in your debt!

**Comment by madebr** [comment]
With MSVC, you can indeed build in a developer console with MSVC + ninja/nmake/.. (that's what I did)
Alternatively, you can generate a MSVC project using the [Visual Studio generators](https://cmake.org/cmake/help/latest/manual/cmake-generators.7.html#visual-studio-generators). A .sln is then generated that can be opened with Visual Studio.

Have you built the dll's in Release mode? I think by default no optimizations are enabled.
To do this, configure the cmake project with [`-DCMAKE_BUILD_TYPE=Release`](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html).

**Comment by dbry** [comment]
Ah, no, I was not using Release mode. That would explain the difference I think. I'll try that.

Is this a good patch for the  `--enable-stdcall-fixup` warnings?
```
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0142262..d1f7eab 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -515,7 +515,7 @@ if(WAVPACK_BUILD_COOLEDIT_PLUGIN)
         PREFIX ""
     )
     if(MINGW)
-        target_link_libraries(cool_wv4 PRIVATE -mwindows -static-libgcc)
+        target_link_libraries(cool_wv4 PRIVATE -mwindows -static-libgcc "-Wl,--enable-stdcall-fixup")
     endif()
 
 endif()
```

**Comment by dbry** [comment]
> Have you built the dll's in Release mode? I think by default no optimizations are enabled. To do this, configure the cmake project with [`-DCMAKE_BUILD_TYPE=Release`](https://cmake.org/cmake/help/latest/variable/CMAKE_BUILD_TYPE.html). 

Okay, verified that Release build generates expected results. The assembly optimizations are still not included in mingw builds, but those only give an additional 20% improvement, so I can live with that. Thanks!

**Comment by sezero** [comment]
>  The assembly optimizations are still not included in mingw builds,

I don't know how are things with a native environment, but with my cross-build
environment cmake has trouble detecting x86 or x64 cpu _and_ an att assembler
for it. That's possibly why asm optimizations are disabled in your environment
too.

**Comment by dbry** [comment]
Well, I was able to get the assembly working (both `i686` and `x86_64`) by simply adding two new definitions (`CMAKE_SYSTEM_PROCESSOR` and `CMAKE_ASM_COMPILER`) to my toolchain file. The gcc C compiler seems to just accept my .S files, although I assume it parses them and passes them to the assembler with appropriate options, although I was not able to get that to work myself (i.e., by using `i686-w64-mingw32-as` from command-line).

Anyway, working toolchain files looks like this:
```
# Sample toolchain file for building for Windows from an Ubuntu Linux system.
#
# Typical usage:
#    *) install cross compiler: `sudo apt-get install mingw-w64`
#    *) cd build
#    *) cmake -DCMAKE_TOOLCHAIN_FILE=~/mingw-w64-i686.cmake ..
# This is free and unencumbered software released into the public domain.

set(CMAKE_SYSTEM_NAME Windows)
set(TOOLCHAIN_PREFIX i686-w64-mingw32)

set(CMAKE_SYSTEM_PROCESSOR i686)

# cross assembler to use
set(CMAKE_ASM_COMPILER ${TOOLCHAIN_PREFIX}-gcc)

# cross compilers to use for C, C++ and Fortran
set(CMAKE_C_COMPILER ${TOOLCHAIN_PREFIX}-gcc)
set(CMAKE_CXX_COMPILER ${TOOLCHAIN_PREFIX}-g++)
set(CMAKE_Fortran_COMPILER ${TOOLCHAIN_PREFIX}-gfortran)
set(CMAKE_RC_COMPILER ${TOOLCHAIN_PREFIX}-windres)

# target environment on the build host system
set(CMAKE_FIND_ROOT_PATH /usr/${TOOLCHAIN_PREFIX})

# modify default behavior of FIND_XXX() commands
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ONLY)
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ONLY)
```

**Comment by madebr** [comment]
For reference, my `mingw32.cmake` cmake toolchain is (MinGW@Linux):
```cmake
set(CMAKE_SYSTEM_NAME Windows)
set(CMAKE_SYSTEM_PROCESSOR x86)

set(CMAKE_SHARED_LINKER_FLAGS "-lssp")

find_program(CMAKE_C_COMPILER NAMES i686-w64-mingw32-gcc)
find_program(CMAKE_CXX_COMPILER NAMES i686-w64-mingw32-g++)

if(NOT CMAKE_C_COMPILER)
	message(FATAL_ERROR "Failed to find CMAKE_C_COMPILER.")
endif()

if(NOT CMAKE_CXX_COMPILER)
	message(FATAL_ERROR "Failed to find CMAKE_CXX_COMPILER.")
endif()

execute_process(COMMAND "${CMAKE_C_COMPILER}" -print-search-dirs
	RESULT_VARIABLE CC_SEARCH_DIRS_RESULT
	OUTPUT_VARIABLE CC_SEARCH_DIRS_OUTPUT)

if(CC_SEARCH_DIRS_RESULT)
	message(FATAL_ERROR "Could not determine search dirs")
endif()

string(REGEX MATCH ".*libraries: (.*).*" CC_SD_LIBS "${CC_SEARCH_DIRS_OUTPUT}")
string(STRIP "${CMAKE_MATCH_1}" CC_SEARCH_DIRS)
string(REPLACE ":" ";" CC_SEARCH_DIRS "${CC_SEARCH_DIRS}")

foreach(CC_SEARCH_DIR ${CC_SEARCH_DIRS})
	if(CC_SEARCH_DIR MATCHES "=.*")
		string(REGEX MATCH "=(.*)" CC_LIB "${CC_SEARCH_DIR}")
		set(CC_SEARCH_DIR "${CMAKE_MATCH_1}")
	endif()
	if(IS_DIRECTORY "${CC_SEARCH_DIR}")
		if(IS_DIRECTORY "${CC_SEARCH_DIR}/../include" OR IS_DIRECTORY "${CC_SEARCH_DIR}/../lib" OR IS_DIRECTORY "${CC_SEARCH_DIR}/../bin")
			list(APPEND CC_ROOTS "${CC_SEARCH_DIR}/..")
		else()
			list(APPEND CC_ROOTS "${CC_SEARCH_DIR}")
		endif()
	endif()
endforeach()

list(APPEND CMAKE_FIND_ROOT_PATH ${CC_ROOTS})

# search for programs in the build host directories
set(CMAKE_FIND_ROOT_PATH_MODE_PROGRAM NEVER)

if(OVERRIDE_FIND_ROOT_PATH_MODE)
	set(_OVERRIDE_FIND_ROOT_PATH_MODE BOTH)
else()
	set(_OVERRIDE_FIND_ROOT_PATH_MODE ONLY)
endif()
# for libraries and headers in the target directories
set(CMAKE_FIND_ROOT_PATH_MODE_INCLUDE ${_OVERRIDE_FIND_ROOT_PATH_MODE})
set(CMAKE_FIND_ROOT_PATH_MODE_LIBRARY ${_OVERRIDE_FIND_ROOT_PATH_MODE})
set(CMAKE_FIND_ROOT_PATH_MODE_PACKAGE ${_OVERRIDE_FIND_ROOT_PATH_MODE})
```

To get it to build the assembly files, I applied the following patch:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index 0142262..d6ccb70 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -118,13 +118,13 @@ if(LARGE_FILES_SUPPORTED)
 endif()
 check_function_exists(fseeko HAVE_FSEEKO)
 
-if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64.*|x86_64.*|AMD64.*")
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "amd64.*|AMD64.*|[xX]86_64.*")
     if(CMAKE_SIZEOF_VOID_P EQUAL 8)
         set(CPU_ASM_X64 1)
     elseif(CMAKE_SIZEOF_VOID_P EQUAL 4)
         set(CPU_ASM_X86 1)
     endif()
-elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "i386.*|i486.*|i586.*|i686.*|i786.*")
+elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "[iI][34567]86.*|[xX]86.*")
     set(CPU_ASM_X86 1)
 elseif(CMAKE_SYSTEM_PROCESSOR MATCHES "arm*")
     if(CMAKE_SIZEOF_VOID_P EQUAL 4)
```

**Comment by sezero** [comment]
Having to define `CMAKE_ASM_COMPILER` in my toolchain file: Sure, OK.
Having to define `CMAKE_SYSTEM_PROCESSOR` in my toolchain file: Ugh..

After all those, ASM-ATT compiler is misdetected:
```
-- Found assembler: /opt/cross_win32/bin/i686-w64-mingw32-gcc
-- Looking for a ASM-ATT compiler
-- Looking for a ASM-ATT compiler - /usr/bin/as
-- The ASM-ATT compiler identification is GNU
-- Found assembler: /usr/bin/as
```

Maybe it's my cmake's fault? (It's version 3.9.6.)

**Comment by madebr** [comment]
About the architecture, I like how mpg123 handles this and used the same approach for SDL:
https://github.com/madebr/mpg123/blob/master-with-github-ci/ports/cmake/cmake/CheckCPUArch.cmake
=> No more need for `CMAKE_SYSTEM_PROCESSOR`

`CMAKE_SYSTEM_NAME` and `CMAKE_SYSTEM_PROCESSOR` are the cmake equivalent of the GNU triplets, so yeah a lot of variability can be expected.

**Comment by madebr** [comment]
GNU does speak the AT&T dialect, so looks correct to me?

**Comment by sezero** [comment]
> GNU as does speak the AT&T dialect, so looks correct to me?

I would have expected dection of `/opt/cross_win32/bin/i686-w64-mingw32-as`
instead of the host system's one.

**Comment by madebr** [comment]
Yes you're right, I was missing setting `CMAKE_ASM-ATT_COMPILER` in my toolchain.
Isn't this the same behavior as autotools?
There, you need to define `CC`, `CXX` and `AS`. If you forget to set one of these, you end up mixing native and crossbuild compilers.

**Comment by sezero** [comment]
> Isn't this the same behavior as autotools? There, you need to define `CC`, `CXX` and `AS`. If you forget to set one of these

If you don't have a badly written configure.ac, all you have to do is using the `--host=CROSS_PREFIX` switch on your `configure` command line, e.g. `./configure --host=i686-w64-mingw32 ....`, and the rest is automatically handled. You _can_ set CC, CXX, AS, etc variables only if you want to override the `CROSS_PREFIX`'ed ones.

**Comment by madebr** [comment]
`CMAKE_ASM_COMPILER` is detected correctly as the cross compiler. I suppose the dialects are causing issues.

**Comment by sezero** [comment]
I am getting the following errors from current git master:
```
CMake Error: CMake can not determine linker language for target: cool_wv4
CMake Error: Cannot determine link language for target "cool_wv4".
CMake Error: CMake can not determine linker language for target: in_wv
CMake Error: Cannot determine link language for target "in_wv".
```

If I apply the following patch fixes, cmake configuration succeeds:
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d1f7eab..69916fa 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -513,6 +513,7 @@ if(WAVPACK_BUILD_COOLEDIT_PLUGIN)
         DEFINE_SYMBOL AUDITION_EXPORTS
         SUFFIX ".flt"
         PREFIX ""
+        LINKER_LANGUAGE "C"
     )
     if(MINGW)
         target_link_libraries(cool_wv4 PRIVATE -mwindows -static-libgcc "-Wl,--enable-stdcall-fixup")
@@ -536,6 +537,7 @@ if(WAVPACK_BUILD_WINAMP_PLUGIN)
     set_target_properties(in_wv PROPERTIES
         DEFINE_SYMBOL WINAMP_EXPORTS
         PREFIX ""
+        LINKER_LANGUAGE "CXX"
     )
     if(MINGW)
         target_link_libraries(in_wv PRIVATE -mwindows -static-libgcc -static-libstdc++)
```

... **BUT** none of the objects of either plugins are built, and it simply
outputs empty dlls: WTH???
```
Scanning dependencies of target cool_wv4
[ 64%] Linking C shared module cool_wv4.flt
[ 64%] Built target cool_wv4
......
Scanning dependencies of target in_wv
[ 89%] Linking CXX shared module in_wv.dll
[ 89%] Built target in_wv
```

**Comment by madebr** [comment]
In general, when you modify your cmake toolchain you need to remove `CMakeCache.txt` and reconfigure from scratch.
CMake caches the initial values, and won't use new ones.
This is probably done to avoid mixing cached values of multiple toolchains.

This error looks like an error with detection of the C/CXX compiler.

**Comment by sezero** [comment]
I build in a separate directory, and at every attempt, I just remove and recreate that directory. The main library dll and the exes are built OK, so I don't think there is some bad detection.

**Comment by madebr** [comment]
Care to share your cmake toolchain + arguments of cmake?

**Comment by sezero** [comment]
> Care to share your cmake toolchain + arguments of cmake?

`cmake -DCMAKE_BUILD_TYPE=Release -DBUILD_SHARED_LIBS=ON -DCMAKE_TOOLCHAIN_FILE=../Toolchain-MinGW32.cmake ../WavPack2`


I found that https://github.com/dbry/WavPack/commit/8ff1f86c577cf66df78d642dae99d5ce041b0e9d (and https://github.com/dbry/WavPack/commit/6b20158b392ea93dae03937311ef2fcc1d74e7a8) are to blame, which add `$<TARGET_OBJECTS:wavpack>` to those plugins' source list and remove explicit linkage to libwavpack. Reverting that (like patch below), I don't need to specify `LINKER_LANGUAGE` and things build properly.
```diff
diff --git a/CMakeLists.txt b/CMakeLists.txt
index d1f7eab..ded1a74 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -504,7 +504,6 @@ if(WAVPACK_BUILD_COOLEDIT_PLUGIN)
         audition/filters.h
         audition/resource.h
         audition/wavpack.rc
-        $<TARGET_OBJECTS:wavpack>
         ${CMAKE_CURRENT_BINARY_DIR}/audition/cool_wv4.def
     )
     target_include_directories(cool_wv4 PRIVATE include)
@@ -514,6 +513,7 @@ if(WAVPACK_BUILD_COOLEDIT_PLUGIN)
         SUFFIX ".flt"
         PREFIX ""
     )
+    target_link_libraries(cool_wv4 PRIVATE wavpack)
     if(MINGW)
         target_link_libraries(cool_wv4 PRIVATE -mwindows -static-libgcc "-Wl,--enable-stdcall-fixup")
     endif()
@@ -529,7 +529,6 @@ if(WAVPACK_BUILD_WINAMP_PLUGIN)
         winamp/wavpack.rc
         winamp/wasabi/wasabi.cpp
         winamp/wasabi/wasabi.h
-        $<TARGET_OBJECTS:wavpack>
     )
     target_include_directories(in_wv PRIVATE include)
     target_compile_definitions(in_wv PRIVATE $<$<BOOL:${MSVC}>:_CRT_SECURE_NO_DEPRECATE>)
@@ -537,6 +536,7 @@ if(WAVPACK_BUILD_WINAMP_PLUGIN)
         DEFINE_SYMBOL WINAMP_EXPORTS
         PREFIX ""
     )
+    target_link_libraries(in_wv PRIVATE wavpack)
     if(MINGW)
         target_link_libraries(in_wv PRIVATE -mwindows -static-libgcc -static-libstdc++)
     endif()
```

**Comment by madebr** [comment]
I see.
Up until [CMake 3.14](https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html), the target specified in `$<TARGET_OBJECTS:xxx>` needed to be an object library.

**Comment by sezero** [comment]
> I see. Up until [CMake 3.14](https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html), the target specified in `$<TARGET_OBJECTS:xxx>` needed to be an object library. So you can revert that commit, knowing that the plug-ins preferably link to a static wavpack library.

@dbry: I suggest applying the above revert patch (or bumping the cmake requirement to something minimum that works with existing cmake'ry.)

**Comment by sezero** [comment]
> I see. Up until [CMake 3.14](https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html), the target specified in `$<TARGET_OBJECTS:xxx>` needed to be an object library.

BTW: Have I ever mentioned that cmake is evil?

**Comment by madebr** [comment]
The frontend is something to have mixed feelings about, but its cross platform backend is really lovable.

**Comment by madebr** [comment]
> @dbry: I suggest applying the above revert patch (or bumping the cmake requirement to something minimum that works with existing cmake'ry.)

FYI, we use 3.0 as minimum required version for SDL2, 3.16 for SDL_mixer.
SDL3 might also make the jump to 3.16.

**Comment by dbry** [comment]
> > I see. Up until [CMake 3.14](https://cmake.org/cmake/help/v3.14/manual/cmake-generator-expressions.7.html), the target specified in `$<TARGET_OBJECTS:xxx>` needed to be an object library. So you can revert that commit, knowing that the plug-ins preferably link to a static wavpack library.
> 
> @dbry: I suggest applying the above revert patch (or bumping the cmake requirement to something minimum that works with existing cmake'ry.)

I verified that this builds and the Cooledit and Winamp plugins run fine (and with the asm mods I made run just as fast as the MSVC versions). Applied. Thanks all!

**Comment by sezero** [comment]
Great.  And I think all issues in this ticket are resolved.

**Comment by sezero** [comment]
Out of curiosity, does the attached plugin builds (using open watcom) work properly?
If they do I can send a patch.
[plugin.zip](https://github.com/dbry/WavPack/files/10050775/plugin.zip)

**Comment by dbry** [comment]
> Out of curiosity, does the attached plugin builds (using open watcom) work properly? If they do I can send a patch.
> 
> [plugin.zip](https://github.com/dbry/WavPack/files/10050779/plugin.zip)

On the Cooledit one the exported symbols have the `@nn` decoration, so I got the same error dialog message as above. If you fix that it may work.

The Winamp one _does_ have the correct symbols but it causes an immediate crash when loaded. Which could mean that the functions have the correct name but wrong calling convention, or something else bad about the code.

The dependencies on both look fine (just `USER32` and `KERNEL32`).

**Comment by sezero** [comment]
> On the Cooledit one the exported symbols have the `@nn` decoration, so I got the same error dialog message as above. If you fix that it may work.

Can you test this new one then? [cool_wv.zip](https://github.com/dbry/WavPack/files/10051016/cool_wv.zip)

> The Winamp one _does_ have the correct symbols but it causes an immediate crash when loaded. Which could mean that the functions have the correct name but wrong calling convention, or something else bad about the code.

Hmph.. Or possibly some c++ thing. Will try understanding later.

**Comment by dbry** [comment]
Yes, the Cooledit filter works now! And it's very fast...I assume it has the asm code?

**Comment by sezero** [comment]
> Yes, the Cooledit filter works now!

Cool

> And it's very fast...I assume it has the asm code?

Yep.

**Comment by sezero** [comment]
> > The Winamp one _does_ have the correct symbols but it causes an immediate crash when loaded. Which could mean that the functions have the correct name but wrong calling convention, or something else bad about the code.
> 
> Hmph.. Or possibly some c++ thing. Will try understanding later.

It's really possible that it's some incompatibility thing with c++ code (whatever that it's used for, I don't know..)

**Comment by sezero** [comment]
> > > The Winamp one _does_ have the correct symbols but it causes an immediate crash when loaded. Which could mean that the functions have the correct name but wrong calling convention, or something else bad about the code.
> > 
> > 
> > Hmph.. Or possibly some c++ thing. Will try understanding later.
> 
> It's really possible that it's some incompatibility thing with c++ code (that `wasabi` thing -- whatever that it's used for, I don't know..)

That `wasabi` thing seems to be only used by the AlbumArt functions, and unless I'm mistaken those AlbumArt functions doesn't seem referenced anywhere.

**Comment by sezero** [comment]
Does this winamp plugin build work? It's built after removing all wasabi stuff which includes the albumart procs, so it's C-only now: if it works then it's watcom + c++ fault. If not, well...
[in_wv.zip](https://github.com/dbry/WavPack/files/10052704/in_wv.zip)

**Comment by dbry** [comment]
> Does this winamp plugin build work? It's built after removing all wasabi stuff which includes the albumart procs, so it's C-only now: if it works then it's watcom + c++ fault. If not, well... [in_wv.zip](https://github.com/dbry/WavPack/files/10052704/in_wv.zip)

That one works, but it doesn't show the album art. I don't understand how (or why) that wasabi stuff does what it does. But it _is_ how album art is displayed, and _is_ set up and working, and was supposed to handle setting and deleting the album art from files too (but that is disabled).

It's funny because my friend who wrote Monkey's Audio recently (6 months ago) was asking me how to add album art to _his_ plugin (because it's not documented anywhere). He ended up copying my wasabi code (which I got when I was working with winamp developers, and they said it was okay for me to redistribute it). They never said I could _modify_ it, and until you came along I hadn't, but I'm sure it's okay now.

Anyway, I looked at the exported symbols and there's nothing different, so I have no idea how winamp _knows_ that the album art stuff is in there. I even created a stripped version of my working plugin so that no other symbols were visible, and it still would load the album art. Would be curious how that works.

**Comment by sezero** [comment]
OK, so it's definitely some incompatibility watcom has with c++ stuff, so I won't bother with it any longer unless I have a clear understanding of where the problem is and how.

**Comment by sezero** [comment]
> OK, so it's definitely some incompatibility watcom has with c++ stuff, so I won't bother with it any longer unless I have a clear understanding of where the problem is and how.

One last time: This one is linked with dead code elimination disabled.
[in_wv_2.zip](https://github.com/dbry/WavPack/files/10053395/in_wv_2.zip)

I'm not that hopeful, but looking at the map file, maybe .. ??..

**Comment by dbry** [comment]
Okay, took a quick look and I see that the winamp plugin calls Wasabi_Init() when it's init() is called. And looking there I can see that the wasabi code is talking to winamp through some sort of Windows IPC (using `sendMessage()`). So that means that the C++ code does not need to be interoperable because it's just using messages to talk to winamp proper, so it might be possible to debug.

But why?  :smiley:

**Comment by dbry** [comment]
> > OK, so it's definitely some incompatibility watcom has with c++ stuff, so I won't bother with it any longer unless I have a clear understanding of where the problem is and how.
> 
> One last time: This one is linked with dead code elimination disabled. [in_wv_2.zip](https://github.com/dbry/WavPack/files/10053395/in_wv_2.zip)
> 
> I'm not that hopeful, but looking at the map file, maybe .. ??..

Instant crash on starting winamp, just like the first one. I see you must have known about `Wasabi_Init()` because you would have had to delete those calls to link when yoiu took out the wasabi part.

Oh, and if you get _really_ curious...  :smile: 

[Winamp Download Page](https://www.winamp.com/player/downloads/)

**Comment by sezero** [comment]
Well, it's possibly addressing the wrong offsets for class members. Will try looking at it in future. I wish wasabi thing had a simple C-only interface, but whatever.
